### ISSUE TITLE
[Bug]Possible incomplete checkpoint after Megatron → HuggingFace conversion for Qwen3-VL-30B-A3B-Instruct

### ISSUE THREAD

**Issue by liyandong001** [NVIDIA-NeMo/Megatron-Bridge#2300]
When converting a Megatron checkpoint of Qwen3-VL-30B-A3B-Instruct to HuggingFace format, the generated artifacts appear to be significantly smaller than expected. The original pretrained model is approximately 58GB, and the produced model.safetensors.index.json reports a large total_size, suggesting that the full model should be present. However, the actual shard files sum up to only ~4.4GB after conversion.

From inspecting the weight names in the index, parameters from both the language model and vision modules are present, which suggests this is not an adapter-only checkpoint. This makes the size discrepancy unexpected.

The conversion process completes without errors, but the resulting model appears incomplete. It is unclear whether only trainable parameters are being saved, parameters are incorrectly filtered during conversion, or if a checkpoint gathering step is missing.

The expected behavior is that the converted HuggingFace checkpoint should contain the full model weights and have a size comparable to the original pretrained model. Any guidance on whether this is expected behavior or a potential issue in the conversion pipeline would be greatly appreciated.

**Comment by liyandong001** [comment]
python /workspace/Megatron-Bridge/examples/conversion/convert_checkpoints.py export \
  --hf-model /pretrained_models/Qwen3-VL-30B-A3B-Instruct \
  --megatron-path /heckpoints/iter_0000002 \
  --hf-path ./Qwen3-VL-30B-A3B-Instruct-iter_0000002

-rw-r--r-- 1 root root 4.1G Feb 11 09:41 model-00001-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:41 model-00002-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:41 model-00003-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00004-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00005-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00006-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00007-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00008-of-00013.safetensors
-rw-r--r-- 1 root root 4.7G Feb 11 09:42 model-00009-of-00013.safetensors

**Comment by yaoyu-33** [comment]
Hi I am a bit confused about the adapter part. Can you tell me a bit more about what checkpoint you are converting back to hf?
Like where is your megatron checkpoint of qwen3 comes from.

**Comment by liyandong001** [comment]
Regarding the adapter / checkpoint conversion: I was using the v0.2.1 codebase when I ran into this issue. The behavior I described happens in that version. I’ve checked the latest version and this specific problem no longer reproduces there, so it seems to have been fixed already in newer code.

There is also a related masking issue in v0.2.1 that I wanted to mention. In that version, the loss mask calculation does not correctly include the <|im_end|> token in the supervised region. Since the mask is built by matching only the assistant text span (without template tokens), <|im_end|> is typically excluded. After shifting the mask to align with next-token prediction, the step that should supervise the model to predict <|im_end|> ends up masked out. As a result, <|im_end|> does not receive loss, and the model may not reliably learn the correct conversation boundary.

In short, the boundary token <|im_end|> should be part of the supervised span for each assistant turn, otherwise the model does not explicitly learn to terminate the response.

Thanks again for your work on this — the latest version already looks much better.

**Comment by liyandong001** [comment]
PreloadedVLMConversationProvider

**Comment by JRD971000** [comment]
Looks like this may no longer reproduce. Could you confirm on latest main? I'll close if it's resolved, otherwise we'll split out the remaining loss-mask note as its own issue.

**Comment by sajadn** [comment]
**1. Undersized checkpoint after Megatron → HF conversion**
 
  As @liyandong001 confirmed above (Feb 12), this no longer reproduces on newer code, so we're
  treating this as resolved per the reporter's confirmation.
 
  **2. `<|im_end|>` excluded from the supervised span (v0.2.1)**                      
 
  Verified fixed on current main, both in code and empirically.                                                                                                            
                                                                                                                                                                           
  - In `src/megatron/bridge/data/conversation_processing.py`, assistant end tokens are included                                                                            
    in the loss span by default (`include_end_tokens_for_roles=("assistant",)`), and even when                                                                             
    the mask comes from an HF chat template with `{% generation %}` blocks (which only covers
    assistant *content*), `build_assistant_loss_mask` ORs the boundary tokens into the mask on                                                                             
    top — so the turn terminator is always supervised.                                
  - Empirical check with the real `Qwen/Qwen3-VL-30B-A3B-Instruct` tokenizer, using the same
    markers the Qwen-VL collator passes:                                              
                                                                                      
    ```python                                                                         
    conv = [{"role": "user", "content": "What is 2+2?"},                              
            {"role": "assistant", "content": "4"}]                                    
    ids = tok.apply_chat_template(conv, tokenize=True)["input_ids"]                                                                                                        
                                                                                      
    cfg = assistant_mask_boundary_config_from_markers(                                
        tok,                                                                                                                                                               
        assistant_start="<|im_start|>assistant\n",                                    
        assistant_end="<|im_end|>\n",                                                 
        assistant_end_fallbacks=("<|im_end|>",),                                      
    )                                                                                 
    mask = build_assistant_loss_mask(conv, ids, tok, boundary_config=cfg)             
    labels, shifted = build_shifted_labels_and_loss_mask(torch.tensor(ids), mask)     
    ```                                                                               
                                                                                      
    Result: the unshifted mask covers exactly `'4'`, `'<|im_end|>'`, `'\n'` (user turn fully                                                                               
    masked), and after the next-token shift the supervised labels are `'4'`, `'<|im_end|>'`,                                                                               
    `'\n'` — i.e. the step that predicts `<|im_end|>` receives loss, so the model is explicitly
    trained to terminate the assistant turn. This was exactly the failure mode described for                                                                               
    v0.2.1.                                                                           
                                                                                                                                                                           
  Both reported problems are resolved on main, so closing. Please reopen if you still see
  either behavior on a current release.
